frontend/src/components/pages/roles/MatchingUsersCard.tsx

```diff
@@ -0,0 +1,267 @@
+ *
+ * As of the Change Date specified in that file, in accordance with
+ * the Business Source License, use of this software will be governed
+ * by the Apache License, Version 2.0
+ */
+
+import { createConnectQueryKey, useMutation, useQuery } from '@connectrpc/connect-query';
+import { useToast } from '@redpanda-data/ui';
+import { useQueryClient } from '@tanstack/react-query';
+import { Check, Plus, Trash2, X } from 'lucide-react';
+import { useState } from 'react';
+import { Button } from 'components/redpanda-ui/components/button';
+import { Card, CardContent, CardHeader } from 'components/redpanda-ui/components/card';
+import {
+  type ListRoleMembersRequest,
+  SecurityService,
+  type UpdateRoleMembershipRequest,
+} from 'protogen/redpanda/api/dataplane/v1/security_pb';
+import {
+  listRoleMembers,
+  updateRoleMembership,
+} from 'protogen/redpanda/api/dataplane/v1/security-SecurityService_connectquery';
+import type { ListUsersRequest } from 'protogen/redpanda/api/dataplane/v1/user_pb';
+import { listUsers } from 'protogen/redpanda/api/dataplane/v1/user-UserService_connectquery';
+import { AutocompleteInput } from '../acls/new-acl/AutocompleteInput';
+
+interface MatchingUsersCardProps {
+  principalType: string;
+  principal: string;
+}
+
+export function MatchingUsersCard({ principalType, principal }: MatchingUsersCardProps) {
+  // Extract role name from principal (format: "RedpandaRole:roleName")
+  const roleName = principalType === 'RedpandaRole' ? principal.split(':')[1] : '';
+
+  // State for adding new user
+  const [isAddingUser, setIsAddingUser] = useState(false);
+  const [newUserName, setNewUserName] = useState('');
+  const [deletingPrincipal, setDeletingPrincipal] = useState<string | null>(null);
+
+  const queryClient = useQueryClient();
+  const toast = useToast();
+
+  // Fetch role members if it's a RedpandaRole
+  const {
+    data: membersData,
+    isLoading,
+    refetch,
+  } = useQuery(
+    listRoleMembers,
+    {
+      roleName,
+      pageSize: 100,
+    } as ListRoleMembersRequest,
+    {
+      enabled: principalType === 'RedpandaRole' && !!roleName,
+    },
+  );
+
+  // Fetch all users for the combobox
+  const { data: usersData } = useQuery(
+    listUsers,
+    {
+      pageSize: 1000, // Get as many users as possible
+    } as ListUsersRequest,
+    {
+      enabled: isAddingUser && principalType === 'RedpandaRole',
+    },
+  );
+
+  // Set up mutation for adding/removing user to/from role
+  const { mutateAsync: updateMembership, isPending: isSubmitting } = useMutation(updateRoleMembership, {
+    onSuccess: async (_, variables) => {
+      // Invalidate and refetch the role members query
+      await queryClient.invalidateQueries({
+        queryKey: createConnectQueryKey({
+          schema: SecurityService.method.listRoleMembers,
+          cardinality: 'finite',
+          input: { roleName },
+        }),
+      });
+      await refetch();
```

Comment:
Refetching is not good - we should invalidate the query and then it can be refetched in due course when the time is right. Having the refetch alongside the invalidate kind of defeats the purpose of having cache stale time. Ideally you just mark the query as invalidated, causing it to load automatically at another time.